src/nucleotide.rs

```diff
@@ -87,21 +207,41 @@ impl TryFrom<u8> for Nucleotide {
 
     #[inline(always)]
     fn try_from(u: u8) -> Result<Self, Self::Error> {
-        if u < 128 {
-            let v = Self::ASCII_PACK_TABLE[u as usize];
-            if v < Self::N_NUCLEOTIDES {
-                // SAFETY: there are only X variants, with assigned numbers, so 0..=(X - 1)
-                //         are valid reprs of this type
-                Ok(unsafe { std::mem::transmute(v) })
-            } else {
-                Err(TranslationError::BadNucleotide(u.into()))
-            }
+        if u >= 128 {
+            return Err(TranslationError::NonAsciiByte(u));
+        }
+
+        let v = ASCII_TO_NUCLEOTIDE[u as usize];
+        if v == 0 {
+            return Err(TranslationError::BadNucleotide(u.into()));
+        }
+
+        if v.count_ones() == 1 {
+            Ok(unsafe { std::mem::transmute(v) })
```

Comment:
We can remove this unsafe by turning `ASCII_TO_NUCLEOTIDE` into `[Option<NucleotideAmbiguous>; 256]`, since there's a niche for `None` to take. With a trivial `impl TryFrom<NucleotideAmbiguous> for Nucleotide` using `match`, it generates better code than the unsafe:

```
example::try_from:
        test    dil, dil
        js      .LBB0_1
        movzx   eax, dil
        lea     rcx, [rip + .L__unnamed_1]
        movzx   eax, byte ptr [rax + rcx]
        ret
.LBB0_1:
        xor     eax, eax
        ret
```

(there's no error construction in this assembly since I stubbed the errors out for `Err(())`, but I don't think that should affect the relative efficiency much)

Reply:
https://rust.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgArKUcAZPAZMADl3ACNMYhAADgBmUgAHVAVCOwYXNw9vROTUgQCg0JYIqLjLTGtbASECJmICDPdPHytMGzSauoICkPDImPjzLsasltr6nqKSgYBKS1QTYmR2DjNMAGpzdBAUAQA3SIIdgBViAE8ASUFUAFJYgCEbjQBBJ9eAJlibr3usYjwDhAnLQBJhSOsXAkzuDlF08GJsABHcFImbfAAibxun2%2B92ImASxAgJmiaK8mJeCRMYXWjHc61CyHoqFsWHWNwA7I8XuteetnuzYuj1howhpxVxSG8%2BetjoLhaKJRopTy%2BU55SKxVxxSqBXyAOIa0XanVYjkU14vbFfH5/AGYIEgoLgyHQ9aw%2Brw2hIlGIskW624/GE4mkjFvKk0uksBkmJmYFl4LDPYp4YAmBYKdlc6V8gV3YWM5ms9ggAVMLMk3UyuUF2PxxNYE7rCvrKu53nqutFhMlnbq1vt1W8w3duPFpOlw2D6K6jvrADqGp7jdL5cr0Wz6pXfZAcpn1b5AFll%2BPe5Oduu25vOduz6v%2By2N4feQAlU8N3dXklb%2BsTpsgNOz7rAA9CB6xUv8QQKPOACaH7/qW%2B4br%2BO4XiAA7AWBEFnP8LBJoEmAwcO6xCAh54AZh16ofeu5AdeL7rAA0uRD57k%2B1G3n%2BFFThxQ6WnqvL3Kxu7IZxHJ3p%2B6FUT%2BXFoQB9FVqB4EMCy/LzgAaiJ6HfjeEncWxMl6ZJiE7Ips7KesqkELK87CmOUkAbpNGOUhfHGQZdHueC2HWRC84ABLaU57kuaZ7EzmFPGPgell%2BfqWIkcEwVrqFcm0ehYmyfp8mlkZUVseZc7molzxoAw5jrFQDBPsgeB4AA%2BkQDUMBlWBNUwYT0BAMzrAAtHc2Dsj8ADyCRVAwdxOLlKZhGmGZmINdzCe8XgAGwYtm3KCes9A2SwJg2QkTDIAA1h1XUbHWuLBKCy3rKtG3kst87rCwJ3EKgDWmPQChgGA6yBEo9Rba9fIQAAJMgCBRLQhCRGI4IQ3sdTwoIICYKohK9QWg2gyRMoysdZ0XfQuJQzDAB0zUVnVjUmAkCSRKISg9XxKQAF6YJtdZCKgbCQyj/yGAQaIPGDhMQSd521Jd5PQ8Q1NfbT9UNSCADuzMVg6vWDpz3PkhqfMC8jqMi2L22S7ynKYuLJE26VMpA4c/0QGEYAcEwHvgjNqbppml4W/Ozv1K77scAQ3ueZOs3zQHe5ByRIeR5A4fIFHvtzf7Zj9onO3J2HHvABnbWYLH2cKGZefzthrBZyYhBnOsaBYMR%2BcVS7qcewwJeueXC2VyAwQW5ZzzgscLrgiQ6yGhAaAsAk9BsIIDJoknHeh13HAsL3iH9/HR4j9hArT%2Bqc/84vmDLzZTFr%2B3wMp27HvELvPH7znICvkf4En8QM/rOfBeS9GA2VgnfJ2G9H7h3Vq/Vc79B4Lm/vydY085SAMvtfRc4JCBWCoOAvkBct5/V4NHZMfsB47CEEg9U09Z7zwwSA0i2CCC4PwbyQhT8OBnFgSWeBOwwH3WwjQv%2BaD6HAJXl/YOkDC4cFOjwmO5D4630EeBQ0qCAFiKvoww%2BUiH4yL2PIsh9d44aSQWPCEKC/50IvuImy9w2GA2kVvBAhiy6KI/gFMxLpLGyg0TYrRK90QOI4eHdAri%2BEgCCSo/k4I1EiL8UAgJNlPG6M7pw8OPtS4RPsdEpwsSfGiP8Zg0xjs%2BTExlp1MmVoSpWheOVSqzwhBOAuBcBqxwRoNWCHIJwfhsAjWOBcdE2AQDDXuGNCaU1M5x0WrEbA91Ho8yFLVVWzVWquVJjrF6tTnh4CAZsAg2xdgMAOPUE45wABin0WCTKye4mCsyqrT1yvjHaBAzhM3WNgYgn0/51h6lsgSMpAw/ECHDIIEAxDqyYGcBQ/p5zVXWAQc4DUqBXIgMwEZUyK69QGg818RETC0AIFNIQFQqDgn%2BQ8zklsrbvQINDKyTAXlW0JliihZZ5R4xGqdCApLaBUEDoxFlbL461iFFynlfKBUJyFVbEVH8uzioedy3lZLc6yslvKweo4lVDRVVKquGrCYNU5Q8r5RIepzAltmC0QKamvHtaVSMVUapIrOCitFJgRkkhxXjfFChCXEtiNNW5xizAUrFkNalwcqBtkFHMpZXB3h6RpZLfEBBFg1XNVAGYec7UBhInShljTmmtPaZ07pvT%2BmDLmT8Ew7M8Bcx5jmAmfJbpBFNUNbNlqjXGwdHsHGuq3pMHpQgdYexqbIsCEQNm0bW2SxVQOzt6xF1WvnYTbNDVB22y7d8nNa6dr5vnA7apBbngIveoEWdLadqEmnbQHukBsTvGjSVd47xwRuo9fzCAa0NAzEpiYBg6tiBMASGzGc1dzQcDmLQTgXheCeA4FoUgqBOCvjMDZBQCwlhXU%2BDwUgRxkMwbmKdbwXBKbRHeJIAAnNEGjXhojRC8FwLwHIfBwY4JIRDmheBoY4LwSuyoiNaDmHAWASB6F0EiOQSgUn6BRGIFwNaXhlRYD2HgZYGk8CYHVmNRgnACM0CJZESubteOkDmswc4hneDz2viNBgtAzgWawO9Iw4hiOkHwPidoBxK5ecxm0Q6KwCPToqBZuGYRQPnBcE2LzSLdm2bmFQAwwAFDad0/ppDRnZAiDEOwKQMhBCKBUOoLzuhJQGCMCAUw5h9B4DCJXSAcxUDjTSAFvqI1BMVDaBNBwDBnCuCaBID9/hCJTH6NqHIKQJojE8JKJIs20iTD6FESUrR2jVHGA0YbWQk3lEqB0Hbq3ihTeVEMeo829CXe6BNtbEg5jYcWMsR7%2Bh4M8a8/x9YdX6XrAoypymGgAG4EINPa0XAZi8BEyR0gZHJAUeiGtDksRlMckkBoRja1WPva459lD/HBMgGE7xsTMBEAgAWAQKkFgKDoOk8QYIrAVi/eQP9ym2Oge8EwPgIg/xtiSn4IIfL4gitC/kEoNQFnKukBA2B5LuOEOkCQwTzgI1Do05QbG1n7PAfA6BBfBn7J8NQ8I6T0jIA1qU1iBoDka0aOxGRxoej0RtQfs49x5XFnCeWGJ2b4jubcfvHx3xzg0PzekBOSkewkggA%3D%3D

Reply:
godbolt link: https://rust.godbolt.org/z/GbdG6z33M

Reply:
That's very nice!